### DISCUSSION TITLE
How do I change env var DOMAIN properly so all attachments still work with the new domain name?

### DISCUSSION THREAD

**Discussion by ElTopo** [dani-garcia/vaultwarden#5206] [Q&A] [answered]
Hello,

If I change **DOMAIN** in docker-compose-yml / services / vaultwarden / environment / DOMAIN from an expiring https://domain1.com/ to the new https://vw.domain2.com/, what else should I do to make sure all current attachments still work with the new domain name?

Thanks.

**Comment by stefan0xC** [comment]
You should make sure that you don't have a [`data/config.json`](https://github.com/dani-garcia/vaultwarden/wiki/Configuration-overview). Otherwise you might want to update the values via the `/admin` page instead.

**Reply by ElTopo** [reply]
Somehow, I do have a data/config.json (in ./vw_data:/data folder, defined in volumes: of docker-compose.yml). 

So, my understand is:
1. If I don't have config.json, I just need to change docker-compose.yml (DOMAIN: line, from the old domain to the new one);
2. if I do have config.json, I need to login /admin page and change "Domain URL" to the new domain name, the config.json will be updated automatically, and current attachments are switched to the new domain.

is it correct?

**Reply by stefan0xC** [reply, marked as answer]
> Somehow, I do have a data/config.json

As explained in the linked wiki page this happens if you configure Vaultwarden via the `/admin` page.

Either way you want to update the `DOMAIN` setting so that this works because Vaultwarden uses it https://github.com/dani-garcia/vaultwarden/blob/96813b1317122e2e24d42dcad6c08e42548db9e4/src/auth.rs#L390-L391 to create the attachment URL in the returned JSON here:
https://github.com/dani-garcia/vaultwarden/blob/96813b1317122e2e24d42dcad6c08e42548db9e4/src/api/core/sends.rs#L568-L572

So if you don't update the domain to the correct base URL and open the site under the new domain anyway, the client will try to fetch the encrypted attachment file from the old location (so even if Vaultwarden were still reachable under the old domain the request would still fail because of a CSP violation).

**Reply by ElTopo** [reply]
@stefan0xC Thanks.

I did read "How to configure Vaultwarden", however it wasn't precise enough for a normal user.
It says:
_1. setting up environment variables,
2. using an ENV_FILE and
3. via a config.json, not recommended (which can be generated and managed via the [admin page](https://github.com/dani-garcia/vaultwarden/wiki/Enabling-admin-page))._

. It does not say "how to set up environment variables", I have to assume "using .env with docker-compose.yml"
. It does not mention that if I specify DOMAIN in .env, I can later just change DOMAIN in .env to a different URL without affecting all my current attachments (I hope so)
. it does not mention that if I don't have DOMAIN in .env, I still can use the admin page to change the domain URL without affecting attachments (I guess the answer is positive but again the doc does not precisely state so)
. It says config.json is "not recommended", but it does not say which 1 of these 3 is recommended or the right way to do it.

At this time, I just specify/change DOMAIN in .env and ignore current config.json (I logged into the admin page and verified the URL is correct in general settings).

**Reply by stefan0xC** [reply]
Well, that page is a general overview and not specific to how everything affects everything else. So I don't know why you'd think that it would explain the question I've answered above? I pointed it out because it discusses the order of precedence so you don't get confused why changing a value in one file wouldn't necessarily affect the outcome.

Given your statements I am afraid that you have not read the page beyond those bulletin points.

> . It does not say "how to set up environment variables", I have to assume "using .env with docker-compose.yml"
> [...]
>  . It says config.json is "not recommended", but it does not say which 1 of these 3 is recommended or the right way to do it.

The page explicitly states that you should familiarize yourself with your setup as explaining every nuance of every possible method would go easily beyond the scope of the page.

> ### Using environment variables
>
> The recommended way to configure Vaultwarden is via environment variables. Depending on how you run Vaultwarden (e.g. directly, in a containerized environment, via systemd, etc.), there are different ways to set the environment variables, so familiarize yourself with your platform and method of installation.
> 
> Most environment variables that can be set are found in the .env.template file. You can also use that file as a basis for an environment file for your container environment (e.g. via an [env_file](https://docs.docker.com/compose/environment-variables/set-environment-variables/#use-the-env_file-attribute) attribute) or with a systemd service (c.f. [EnvironmentFile=](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#EnvironmentFile=)) – just don't confuse this file with [the ENV_FILE method](https://github.com/dani-garcia/vaultwarden/wiki/Configuration-overview#using-an-env_file) below!

Also if we were to explain [how to setup environment variables on a general purpose Linux system](https://wiki.archlinux.org/title/Environment_variables) this would be more than misleading for the average user as they would most likely be using a container image and not run the process directly.

> At this time, I just specify/change DOMAIN in .env and ignore current config.json (I logged into the admin page and verified the URL is correct in general settings).

Okay. even if I'm not sure what you mean by "ignore current config.json" but if it works that's great.

**Reply by ElTopo** [reply]
Well I felt like that the "How to" page needs some improvement; it might be good for developers (who can go ahead reading source code to get answers) but left questions to normal users.

By "ignore current config.json", I meant that I would not muck around with the .json file from now on, and simply use .env file and /admin page.

**Reply by auspex** [reply]
It's worth noting that when using `docker compose`, the `.env` file _should_ be the recommended way, as you run into the requirement to escape the $ characters in your admin key if you set environment variables directly in the compose file.
